In [ ]:
csv_path ="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"

In [ ]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
 "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
 "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
 "peak-rpm","city-mpg","highway-mpg","price"]

In [ ]:
#Membaca data
import pandas as pd
df=pd.read_csv(csv_path,names=headers)
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [ ]:
#Jumlah Data
print(df.shape)

(205, 26)


In [ ]:
#Cek tipe data
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Jumlah kolom dengan type data int64 : 5 Kolom

Jumlah kolom dengan type data float64 : 5 Kolom

**Jawaban Nomor 2**

Jumlah nilai unik (unique) dari fuel-type adalah 2, Kolom "fuel-type" memiliki nilai unik **"gas"** atau **"diesel"**

In [ ]:
#mengkonversi ? ke nan
import numpy as np
df.replace("?", np.nan, inplace=True)
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [ ]:
missing_data = df.isnull()
missing_data.head()

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

In [ ]:
#menghitung missing value setiap kolom
for column in missing_data.columns.values.tolist():
 print(column)
 print(missing_data[column].value_counts())
 print("")

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

make
False    205
Name: make, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system
Fa

In [ ]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

Praproses Data

**Replace dengan mean:**
1. "normalized-losses": 41 missing data, replace dengan mean
2. "stroke": 4 missing data, replace dengan mean
3. "bore": 4 missing data, replace dengan mean
4. "horsepower": 2 missing data, replace dengan mean
5. "peak-rpm": 2 missing data, replace dengan mean

**Replace dengan frequency:**
1. "num-of-doors": 2 missing data, replace dengan "four".
Alasan: kebanyakan memiliki 4 pintu, maka kita replace dengan 4.

**Drop baris:**
1. "price": 4 missing data, delete seluruh baris.

In [ ]:
#Hitung rata-rata Kolom 'normalized-loses'
avg_norm_loss= df["normalized-losses"].astype("float").mean(axis=0)
print("Average of Normalized-losses : ", avg_norm_loss)

Average of Normalized-losses :  122.0


In [ ]:
#Stroke
avg_stroke= df["stroke"].astype("float").mean(axis=0)
print("Average of stroke : ", avg_stroke)

Average of stroke :  3.2554228855721337


In [ ]:
#bore
avg_bore= df["bore"].astype("float").mean(axis=0)
print("Average of bore : ", avg_bore)

Average of bore :  3.3297512437810957


In [ ]:
#horsepower
avg_horsepower= df["horsepower"].astype("float").mean(axis=0)
print("Average of horsepower : ", avg_horsepower)

Average of horsepower :  104.25615763546799


In [ ]:
#peak rpm
avg_peakrpm= df["peak-rpm"].astype("float").mean(axis=0)
print("Average of peak-rpm : ", avg_peakrpm)

Average of peak-rpm :  5125.369458128079


Replace NAN

In [ ]:
df["normalized-losses"].replace(np.nan, avg_norm_loss, inplace=True)
df["stroke"].replace(np.nan, avg_stroke, inplace = True)
df["bore"].replace(np.nan, avg_bore, inplace=True)
df["horsepower"].replace(np.nan, avg_horsepower, inplace=True)
df["peak-rpm"].replace(np.nan, avg_peakrpm, inplace=True)

In [ ]:
#Pada kasus : "num-of-doors": 2 missing data, replace dengan "four"
df['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

In [ ]:
#Normalisasi "num-of-doors": 2 missing data, replace dengan "four"
df['num-of-doors'].value_counts().idxmax()
df["num-of-doors"].replace(np.nan, "four", inplace=True)

In [ ]:
#Terakhir, hapus semua baris yag tidak memiliki nilai pada kolom price:
# drop baris dengan NaN dalam kolom "price"
df.dropna(subset=["price"], axis=0, inplace=True)
# reset index, karena kita menge-drop 2 baris
df.reset_index(drop=True, inplace=True)

In [ ]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [ ]:
#Nilai mean (rata-rata) highway-mpg setelah di praproses adalah
avg_highwaympg= df["highway-mpg"].astype("float").mean(axis=0)
print("Average of Highway-mpg : ", avg_highwaympg)

Average of Highway-mpg :  30.686567164179106


In [ ]:
df.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

Perbaiki format data

In [ ]:
df[["bore", "stroke"]] = df[["bore", "stroke"]].astype("float")
df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")
df[["price"]] = df[["price"]].astype("float")
df[["peak-rpm"]] = df[["peak-rpm"]].astype("float")

In [ ]:
df.dtypes

symboling              int64
normalized-losses      int64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower            object
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

**Regresi**

**No 4**

In [ ]:
#load modul linear regression
from sklearn.linear_model import LinearRegression
lm=LinearRegression()
lm

LinearRegression()

In [ ]:
X = df[['highway-mpg']]
Y = df['price']

In [ ]:
#fit model linear dengan highway-L/100km
lm.fit(X,Y)

LinearRegression()

In [ ]:
#Berapakah nilai intercept (a)?
lm.intercept_

38423.3058581574

In [ ]:
#berapakan nilai Slope atau Gradien (b)?
lm.coef_

array([-821.73337832])

**price** = 38423.3058581574 - 821.73337832 x highway-mpg

MSE Highway-mpg terhadap Price

In [ ]:
from sklearn.metrics import mean_squared_error
mean_squared_error(df['highway-mpg'],df['price'])

236536299.00995025

R^*2* Data Highway-mpg terhadap Price

In [ ]:
from sklearn.metrics import r2_score
r2_score(df['highway-mpg'],df['price'])


-5118149.227782062

In [ ]:
nilai_price1 = lm.intercept_ + (lm.coef_ *25)
nilai_price1

array([17879.97140011])

In [ ]:
nilai_price2 = lm.intercept_ + (lm.coef_ *28)
nilai_price2

array([15414.77126514])

Prediksi harga jika highway-mpg=25

price = 17879.97140011

Prediksi harga jika highway-mpg=28

Price = 15414.77126514

**No 5**

In [ ]:
from sklearn.preprocessing import MinMaxScaler 
scaler = MinMaxScaler(feature_range = (0,1)) 
Fitur = scaler.fit_transform(df[["symboling","normalized-losses","wheel-base","length","width","height","curb-weight","engine-size","bore","stroke","compression-ratio","peak-rpm","city-mpg","highway-mpg"]])
print(Fitur)
Label = df['price']
print(Label)

[[1.         0.29842932 0.05830904 ... 0.34693878 0.22222222 0.28947368]
 [1.         0.29842932 0.05830904 ... 0.34693878 0.22222222 0.28947368]
 [0.6        0.29842932 0.2303207  ... 0.34693878 0.16666667 0.26315789]
 ...
 [0.2        0.15706806 0.65597668 ... 0.55102041 0.13888889 0.18421053]
 [0.2        0.15706806 0.65597668 ... 0.26530612 0.36111111 0.28947368]
 [0.2        0.15706806 0.65597668 ... 0.51020408 0.16666667 0.23684211]]
0      13495.0
1      16500.0
2      16500.0
3      13950.0
4      17450.0
        ...   
196    16845.0
197    19045.0
198    21485.0
199    22470.0
200    22625.0
Name: price, Length: 201, dtype: float64


In [ ]:
xlatih = Fitur[:-50,:]
ylatih = Label[:-50]

xuji = Fitur[-50:,:]
yuji = Label[-50:]

In [ ]:
from sklearn.neural_network import MLPRegressor
model = MLPRegressor(activation='logistic', solver='lbfgs', hidden_layer_sizes=(10,10), random_state=0)
model.fit(xlatih, ylatih)
print(model.score(xlatih,ylatih))

-9.748477580728832e-10


In [ ]:
prediksi=model.predict(xuji)
print(prediksi)
print(yuji)

[13587.08808535 13587.08807477 13587.08807438 13587.08807593
 13587.08809126 13587.08809406 13587.08806875 13587.08806884
 13587.08807597 13587.08807613 13587.08808176 13587.08808192
 13587.08808667 13587.08808665 13587.08808671 13587.08808727
 13587.08808742 13587.08809096 13587.08808077 13587.08807076
 13587.08807412 13587.08807672 13587.08807432 13587.08808821
 13587.08808726 13587.08806605 13587.08806451 13587.08811165
 13587.0881011  13587.0881117  13587.08810115 13587.08810141
 13587.0881065  13587.0880997  13587.08809933 13587.08808899
 13587.08807452 13587.08808766 13587.08808728 13587.08806807
 13587.0880792  13587.08806942 13587.08808041 13587.08805277
 13587.08806467 13587.08807648 13587.08806847 13587.08806111
 13587.0880728  13587.08806888]
151     8778.0
152     6938.0
153     7198.0
154     7898.0
155     7788.0
156     7738.0
157     8358.0
158     9258.0
159     8058.0
160     8238.0
161     9298.0
162     9538.0
163     8449.0
164     9639.0
165     9989.0
166    1119

Evaluasi Uji 

In [ ]:
from sklearn.metrics import mean_squared_error
mean_squared_error(yuji, prediksi)

20788239.592959844

In [ ]:
from sklearn.metrics import r2_score
r2_score(yuji,prediksi)

-0.12642365661797084